Megaparsec/Megaparsec.lean

```diff
@@ -169,58 +152,117 @@ class MonadParsec (M: Type -> Type) [Monad M] [Alternative M] where
   -- | Update parser state with @phi@.
-  updateParserState (phi: (State S E -> State S E)): M Unit
-
--- #eval (4 : Id Nat) + (4 : Nat)
-#eval (4 : Id Nat)
-#eval (4 : Nat)
-#eval (pure 4 : Id Nat)
-#eval ((pure 4 : Id Nat) : Nat)
-#check liftM
-
--- instance: Monad (MonadParsec M) where
-
--- instance: MonadLiftT (MonadParsec Mᵤ) Mᵥ where
---   monadLift {α: Type} (???):
-
-instance [Monad M] [Alternative M] [mₚ: MonadParsec M]: MonadParsec (StateT σ M) where
-  E := mₚ.E
-  S := mₚ.S
-  stream := mₚ.stream
-  -- parseError a err := pure $ liftM a (mₚ.parseError a err)
-  parseError a err := sorry
-  label := sorry
-  attempt := sorry
-  lookAhead := sorry
-  notFollowedBy := sorry
-  withRecovery := sorry
-  observing := sorry
-  eof := sorry
-  token := sorry
-  tokens := sorry
-  takeWhileP := sorry
-  takeWhile1P := sorry
-  takeP := sorry
-  getParserState := sorry
-  updateParserState := sorry
--- instance (W: Type -> Type) [Monoid W] (M: Type -> Type) [mₚ: MonadParsec M]: MonadParsec (RWST)
+  updateParserState (phi: (State S E → State S E)): M Unit
+
+
+-- util
+
+def fixs {A B C : Type} (c : C) : Either A (B × C) → (Either A B) × C
+  | .left a => ⟨ .left a, c ⟩
+  | .right ⟨ a, b ⟩ => ⟨ .right a, c ⟩
+
+def fixs' {A B C W : Type} [m : Monoid W] (c : C) : Either A (B × C × W) → (Either A B) × C × W
+  | .left a => ⟨ .left a, c, m.one ⟩
+  | .right ⟨ a, b, w ⟩ => ⟨ .right a, c, w ⟩
+
+def updatePair {A B C : Type} (c : C) : A × B → A × C
+  | ⟨ a, _ ⟩ => ⟨ a, c ⟩
+
+def fst {A B : Type} : A × B → A
+  | ⟨ a, _⟩ => a 
+
+def seqComp {A B : Type} {M : Type → Type} [Monad M] (ma : M A) (mb : M B) :=
+  ma >>= λ _ => mb
+--
+
+instance (E S : Type) [Monad M] [Alternative M] [Stream S] [mₚ: MonadParsec E S M] : MonadParsec E S (StateT σ M) where
+  parseError A err := liftM (mₚ.parseError A err)
+  label A f st := (mₚ.label E S (A × σ) f) ∘ st
+  attempt A st := mₚ.attempt E S (A × σ) ∘ st
+  lookAhead A state x := mₚ.lookAhead E S (A × σ) (state x)
+  notFollowedBy A state x := seqComp (mₚ.notFollowedBy E S A (fst <$> state x)) (pure (Unit.unit , x))
+  withRecovery A cont state x := mₚ.withRecovery (A × σ) (fun e => (cont e) x) (state x)
+  observing A f x := fixs x <$> (mₚ.observing (A × σ) (f x))
+  eof := liftM mₚ.eof
+  token A test mt := liftM (mₚ.token E A test mt)
+  tokens A e ts := liftM (mₚ.tokens E S A e ts)
+  takeWhileP A ms p := liftM (mₚ.takeWhileP E S A ms p)
+  takeWhile1P A ms p := liftM (mₚ.takeWhile1P E S A ms p)
+  takeP A l n := liftM (mₚ.takeP E S A l n)
+  getParserState := liftM mₚ.getParserState
+  updateParserState f := liftM (mₚ.updateParserState f)
+
+-- | RWS monad and its transformer and 
+
+def void {A : Type} {M : Type → Type} [Monad M] (ma : M A) : M Unit :=
+  (λ _ => Unit.unit) <$> ma
+
+def RWST (R W S : Type u) (M : Type u → Type v) (A : Type u) : Type (max u v) :=
+  R → S → M (A × S × W)
+
+instance (R W S : Type) (M : Type → Type) [m : Monoid W]  [Monad M] : Monad (RWST R W S M) where
```

Comment:
Are you sure it'll automatically derive Pure, Bind, Functor etc? I like Lean's granular type classes, but maybe it also automatically derive stuff up the class hierarchy.